README
NAME
    Net::Amazon::Config - Manage Amazon Web Services credentials

VERSION
    This documentation describes version 0.001.

SYNOPSIS
  Example
         use Net::Amazon::Config;
 
         # default location and profile
         my $profile = Net::Amazon::Config->new->get_profile;
 
         # use access key ID and secret access key with S3 
         use Net::Amazon::S3;
         my $s3 = Net::Amazon::S3->new(
           aws_access_key_id     => $profile->access_key_id,
           aws_secret_access_key => $profile->secret_access_key,
         );

  Config Format
       default = johndoe
       [johndoe]
       access_key_id = XXXXXXXXXXXXXXXXXXXX
       secret_access_key = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
       certificate_file = my-cert.pem
       private_key_file = my-key.pem
       ec2_keypair_name = my-ec2-keypair
       ec2_keypair_file = ec2-private-key.pem
       aws_account_id = 0123-4567-8901
       canonical_user_id = <64-character string>

DESCRIPTION
    This module lets you keep Amazon Web Services credentials in a
    configuration file for use with different tools that need them.

USAGE
  new()
       my $config = Net::Amazon::Config->new( %params );

    Valid %params entries include:

    *   config_dir -- directory containing the config file (and the default
        location for other files named in the config file). Defaults to
        "$HOME/.amazon"

    *   config_file -- defaults to "profiles.conf"

    Returns an object or undef if no config file can be found.

  config_path()
       my $path = $config->config_path;

    Returns the absolute path to the configuration file.

  get_profile()
       my $profile = $config->get_profile( $name );

    If $name is omitted or undefined, returns the profile named in the
    top-level key "default" in the config file. If the profile does not
    exist, get profile returns undef or an empty list.

ENVIRONMENT
    *   NET_AMAZON_CONFIG -- absolute path to config file or file name
        relative to the configuration directory

    *   NET_AMAZON_CONFIG_DIR -- configuration directory

BUGS
    Please report any bugs or feature requests using the CPAN Request
    Tracker web interface at
    <http://rt.cpan.org/Dist/Display.html?Queue=Net-Amazon-Config>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

SEE ALSO
    *   About AWS Security Credentials: http://tinyurl.com/yh93cjg

AUTHOR
    David A. Golden (DAGOLDEN)

COPYRIGHT AND LICENSE
    Copyright (c) 2010 by David A. Golden. All rights reserved.

    Licensed under Apache License, Version 2.0 (the "License"). You may not
    use this file except in compliance with the License. A copy of the
    License was distributed with this file or you may obtain a copy of the
    License from http://www.apache.org/licenses/LICENSE-2.0

    Files produced as output though the use of this software, shall not be
    considered Derivative Works, but shall be considered the original work
    of the Licensor.

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



Changes
Revision history for Net-Amazon-Config

{{$NEXT}}

0.002     2014-02-27 22:29:04-05:00 America/New_York

    [CHANGED]

    - Updated metadata, repo and bugtracker

    - Switched distribution management to Dist::Zilla

0.001 Thu Mar  4 17:03:50 EST 2010
    - first release to CPAN


Build.PL
# Copyright (c) 2010 by David Golden. All rights reserved.
# Licensed under Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License was distributed with this file or you may obtain a 
# copy of the License from http://www.apache.org/licenses/LICENSE-2.0

use strict;
use lib 'inc';
eval "require Pod::WikiDoc";
my $class = $@ ? "Module::Build" : "Module::Build::WikiDoc";
eval "require $class";
my $build = $class->new( 
    module_name => 'Net::Amazon::Config',
    dist_author => 'David Golden <dagolden@cpan.org>',
    license => 'apache',
    create_readme => 1,
    create_makefile_pl  => 'traditional',
    requires => {
        'perl' => 0, # not specified
        'Carp' => 0,
        'Config::Tiny' => 2.12,
        'constant' => 1, # 1.00 in perl 5.00504; 1.02 in perl 5.006002; 1.17 in perl 5.008009; 1.18 is new
        'Object::Tiny' => 1.06,
        'Params::Validate' => 0.91,
        'Path::Class' => 0.17,
    },
    build_requires => {
        'File::Find' => 0,
        'File::Spec' => 3.31,
        'Test::More' => 0.88,
    },
    meta_add            => { 
        no_index => { 
            directory => [ qw/ examples inc t xt /],
        },
        resources => {
          repository => "http://github.com/dagolden/Net-Amazon-Config/", 
          bugtracker => "http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-Amazon-Config",
        },
    },
);
$build->create_build_script;



Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3603
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'Net::Amazon::Config',
          'VERSION_FROM' => 'lib/Net/Amazon/Config.pm',
          'PREREQ_PM' => {
                           'Carp' => 0,
                           'Config::Tiny' => '2.12',
                           'File::Find' => 0,
                           'File::Spec' => '3.31',
                           'Object::Tiny' => '1.06',
                           'Params::Validate' => '0.91',
                           'Path::Class' => '0.17',
                           'Test::More' => '0.88',
                           'constant' => 1
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [],
          'PL_FILES' => {}
        )
;


lib/Net/Amazon/Config/Profile.pm
use strict;
use warnings;

package Net::Amazon::Config::Profile;
# ABSTRACT: Amazon credentials for given profile
# VERSION

use Params::Validate ();

my @attributes;

BEGIN {
    @attributes = qw(
      profile_name
      access_key_id
      secret_access_key
      certificate_file
      private_key_file
      ec2_keypair_name
      ec2_keypair_file
      cf_keypair_id
      cf_private_key_file
      aws_account_id
      canonical_user_id
    );
}

use Object::Tiny @attributes;

sub new {
    my ( $class, $first, @rest ) = @_;
    my @args = ref $first eq 'ARRAY' ? (@$first) : ( $first, @rest );
    my %args = Params::Validate::validate( @args, { map { $_ => 0 } @attributes } );
    return bless \%args, $class;
}

1;

__END__

=begin wikidoc

= DESCRIPTION

This module defines a simple object representing a 'profile' of
Amazon Web Services credentials and associated information.

= USAGE

A profile object is created by [Net::Amazon::Config] based on information
in a configuration file.  The object has the following read-only accessors:

* profile_name -- as provided in the configuration file
* access_key_id -- identifier for REST requests
* secret_access_key -- used to sign REST requests
* certificate_file -- path to a file containing identifier for SOAP requests
* private_key_file -- path to a file containing the key used to sign SOAP requests
* ec2_keypair_name -- the name used to identify a keypair when launching an EC2 instance
* ec2_keypair_file -- the private key file used by ssh to connect to an EC2 instance
* cf_keypair_id -- identifier for CloudFront requests
* cf_private_key_file -- path to a file containing the key use to sign CloudFront requests
* aws_account_id -- identifier to share resources (except S3) 
* canonical_user_id -- identifier to share resources (S3 only)

If an attribute is not set in the configuration file, the accessor will
return undef.

= SEE ALSO

* [Net::Amazon::Config]

=end wikidoc

=cut



lib/Net/Amazon/Config.pm
use strict;
use warnings;

package Net::Amazon::Config;
# ABSTRACT: Manage Amazon Web Services credentials
# VERSION

use Carp ();
use Config::Tiny 2.12 ();
use Net::Amazon::Config::Profile ();
use Params::Validate 0.91 ();
use Path::Class 0.17      ();
use Object::Tiny 1.06 qw(
  config_dir
  config_file
  config_path
);

use constant IS_WIN32 => $^O eq 'MSWin32';

sub _default_dir {
    my $base = Path::Class::dir( IS_WIN32 ? $ENV{USERPROFILE} : $ENV{HOME} );
    return $base->subdir('.amazon')->absolute->stringify;
}

sub new {
    my $class = shift;
    my %args  = Params::Validate::validate(
        @_,
        {
            config_dir  => { default => $ENV{NET_AMAZON_CONFIG_DIR} || _default_dir, },
            config_file => { default => $ENV{NET_AMAZON_CONFIG}     || 'profiles.conf', }
        }
    );

    if ( Path::Class::file( $args{config_file} )->is_absolute ) {
        $args{config_path} = $args{config_file};
    }
    else {
        $args{config_path} =
          Path::Class::dir( $args{config_dir} )->file( $args{config_file} );
    }

    unless ( -r $args{config_path} ) {
        die "Could not find readable file $args{config_path}";
    }

    return bless \%args, $class;
}

sub get_profile {
    my ( $self, $profile_name ) = @_;
    my $config = Config::Tiny->read( $self->config_path );

    $profile_name = $config->{_}{default} unless defined $profile_name;
    my $params = $config->{$profile_name}
      or return;

    $params->{profile_name} = $profile_name;
    my $profile = eval { Net::Amazon::Config::Profile->new($params) };
    if ($@) {
        Carp::croak "Invalid profile: $@";
    }
    return $profile;
}

1;

__END__

=begin wikidoc

= SYNOPSIS

== Example

    use Net::Amazon::Config;
    
    # default location and profile
    my $profile = Net::Amazon::Config->new->get_profile;

    # use access key ID and secret access key with S3 
    use Net::Amazon::S3;
    my $s3 = Net::Amazon::S3->new(
      aws_access_key_id     => $profile->access_key_id,
      aws_secret_access_key => $profile->secret_access_key,
    );
    
== Config Format

  default = johndoe
  [johndoe]
  access_key_id = XXXXXXXXXXXXXXXXXXXX
  secret_access_key = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
  certificate_file = my-cert.pem
  private_key_file = my-key.pem
  ec2_keypair_name = my-ec2-keypair
  ec2_keypair_file = ec2-private-key.pem
  aws_account_id = 0123-4567-8901
  canonical_user_id = <64-character string>
  
= DESCRIPTION

This module lets you keep Amazon Web Services credentials in a
configuration file for use with different tools that need them.

= USAGE

== new() 

  my $config = Net::Amazon::Config->new( %params );

Valid {%params} entries include:

* config_dir -- directory containing the config file
(and the default location for other files named in the config file).  
Defaults to {$HOME/.amazon}
* config_file -- defaults to {profiles.conf}

Returns an object or undef if no config file can be found.

== config_path()

  my $path = $config->config_path;

Returns the absolute path to the configuration file.

== get_profile()

  my $profile = $config->get_profile( $name );

If {$name} is omitted or undefined, returns the profile named in the
top-level key {default} in the config file. If the profile does not
exist, get profile returns undef or an empty list.

= ENVIRONMENT

* NET_AMAZON_CONFIG -- absolute path to config file or file name relative
to the configuration directory
* NET_AMAZON_CONFIG_DIR -- configuration directory 

= SEE ALSO

* About AWS Security Credentials: http://tinyurl.com/yh93cjg

=end wikidoc

=cut

